### PR TITLE
[cupertino] improve cupertino picker performance by using at most one opacity layer

### LINKED ISSUE #124703: [Cupertino] ListWheelViewport opacity optimizations.
Currently the cupertino picker will wrap built items in opacity as they scroll offscreen. When applied to text, this can be quite inefficient as we need to create additional offscreen surfaces to perform the opacity op. While the engine (both Skia and Impeller) does have a form of opacity optimization, its generally hard to apply to text.

Instead, I believe the cupertino picker's item builder callback should expose the current opacity. This would allow the caller to determine if they wanted to apply an opacity saveLayer or to directly apply the opacity to the paint color of the items. For text this would usually be correct and lead to substantially better performance.

https://github.com/flutter/flutter/issues/124658
https://github.com/flutter/flutter/issues/121476

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/124703

Rather than applying an opacity layer per picker item, apply them to all partially opaque picker items at once. This dramatically improves performance for impeller, by reducing the number of required subpasses and texture allocations signficantly.

Before:
Doesn't finish, 100s of passes

After:

image

image

See also:

https://github.com/flutter/flutter/issues/124658

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Letting this bake to see if any existing tests fail 😄

**Comment by jonahwilliams** [comment]
Previously with a cupertino picker containing ~N items, we would end up with N opacity layers. Now we only have 1 for all offcenter widgets and potentially 2 more for partially offcenter items - max of 3.

**Comment by jonahwilliams** [comment]
Actually we can go further and combine all partially opaque shapes together, this gives us a single opacity layer regardless of the number of layers.

**Comment by jonahwilliams** [comment]
image

**Comment by jonahwilliams** [comment]
I've looked through the goldens and in general they look OK to me. The change in color is easily explained by different behavior for clipping then opacity versus opacity then clipping. The only concern would be the scale of g3 impact.

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart [hunk 1]
```diff
@@ -828,22 +828,51 @@ class RenderListWheelViewport
...
+
+    // In order to reduce the number of opacity layers, we first paint all
+    // partially opaque children, then finally paint the fully opaque children.
+    context.pushOpacity(offset, (overAndUnderCenterOpacity * 255).round(), (PaintingContext context, Offset offset) {
+      RenderBox? childToPaint = firstChild;
+      while (childToPaint != null) {
+        final ListWheelParentData childParentData = childToPaint.parentData! as ListWheelParentData;
+        _paintTransformedChild(childToPaint, context, offset, childParentData.offset, center: false);
+        childToPaint = childAfter(childToPaint);
+      }
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/list_wheel_viewport.dart:850]: Wondering if we can refactor this into a method (with `center` as a configurable argument) and call that here, above, and below to de-dup this a little and show that it is essentially doing the same thing?

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart [hunk 2]
```diff
@@ -828,22 +828,51 @@ class RenderListWheelViewport
...
   }
 
-  /// Takes in a child with a **scrollable layout offset** and paints it in the
-  /// **transformed cylindrical space viewport painting coordinates**.
+  // Takes in a child with a **scrollable layout offset** and paints it in the
+  // **transformed cylindrical space viewport painting coordinates**.
+  //
+  // If [center] is true, only children that overlap with the center section
+  // will be painted. If [center] is false, only children that do not overlap
+  // will be painted. Opacity is only applied to the partially overlapping
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/list_wheel_viewport.dart:866]: Is this comment about Opacity here actually true? My understanding was that opacity is fully controlled from outside this method now? I.e. it is up to the caller to apply opacity where necessary.

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart [hunk 3]
```diff
@@ -828,22 +828,51 @@ class RenderListWheelViewport
 
   /// Paints all children visible in the current viewport.
   void _paintVisibleChildren(PaintingContext context, Offset offset) {
+    if (overAndUnderCenterOpacity >= 1) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/list_wheel_viewport.dart:831]: should we also go down this code path if useMagnifier is false? Seems like in that case the other code path would paint some children twice since _paintTransformedChild ignores the value of center in that case?

Comment (goderbauer) [reply packages/flutter/lib/src/rendering/list_wheel_viewport.dart:831]: see line 907

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart [hunk 4]
```diff
@@ -876,7 +905,7 @@ class RenderListWheelViewport
 
     final bool shouldApplyOffCenterDim = overAndUnderCenterOpacity < 1;
     if (useMagnifier || shouldApplyOffCenterDim) {
-      _paintChildWithMagnifier(context, offset, child, transform, offsetToCenter, untransformedPaintingCoordinates);
+      _paintChildWithMagnifier(context, offset, child, transform, offsetToCenter, untransformedPaintingCoordinates, center: center);
     } else {
       _paintChildCylindrically(context, offset, child, transform, offsetToCenter);
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/list_wheel_viewport.dart:910]: should we assert that center is null if we go down the else branch?

Actually, the fact that this branch ignores center silently makes the code kinda hard to follow (and I think it is also the source of a bug in the current implementation, see my comment about useMagnifier above). Maybe we need to refactor this somehow to make it easier to follow.

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart [hunk 5]
```diff
@@ -828,22 +828,51 @@ class RenderListWheelViewport
...
-  /// Takes in a child with a **scrollable layout offset** and paints it in the
-  /// **transformed cylindrical space viewport painting coordinates**.
+  // Takes in a child with a **scrollable layout offset** and paints it in the
+  // **transformed cylindrical space viewport painting coordinates**.
+  //
+  // If [center] is true, only children that overlap with the center section
+  // will be painted. If [center] is false, only children that do not overlap
+  // will be painted. Opacity is only applied to the partially overlapping
+  // children painted when [center] is true.
+  // If center is `null` this logic is ignored entirely.
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/list_wheel_viewport.dart:868]: Maybe clearer: if center is null all children will be painted?

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/rendering/list_wheel_viewport.dart:833]: No, the magnifier cannot be disabled if there is opacity.

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/rendering/list_wheel_viewport.dart]: Updated

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/rendering/list_wheel_viewport.dart]: Done

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart (hunk 3 above)
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/list_wheel_viewport.dart:833]: Is that enforced anywhere in the API that `useMagnifier` must be true if `overAndUnderCenterOpacity` is less than 1? Or is it if there is opacity the value of `useMagnifier` is just (silently) ignored?

This is confusing, but I guess not a problem of this PR...

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart [hunk 6]
```diff
@@ -828,22 +828,42 @@ class RenderListWheelViewport
 
   /// Paints all children visible in the current viewport.
   void _paintVisibleChildren(PaintingContext context, Offset offset) {
+    // Note that the magnifier cannot be turned off if the opacity is less than
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/list_wheel_viewport.dart:831]: nit: per style guide just remove the "note that" at the beginning here.

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart (hunk 4 above)
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/list_wheel_viewport.dart:901]: > should we assert that center is null if we go down the else branch?

What about this one? Since this is already pretty confusing it may be worthwhile to put a guardrail in place for when this gets refactored to make the caller instantly aware of the fact that their specification of center is not doing what they may think it is doing?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart (hunk 6 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/rendering/list_wheel_viewport.dart:831]: Done

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/rendering/list_wheel_viewport.dart:832]: the value of use magnifier is just ignored

File: packages/flutter/lib/src/rendering/list_wheel_viewport.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/rendering/list_wheel_viewport.dart:901]: I added the assert. I'm not in love with this code but I don't see a nice way to re-arrange it all...
